### DISCUSSION TITLE
Compiling Examples on Raspbian: undefined reference to `__atomic_load_8'

### DISCUSSION THREAD

**Discussion by brunorzn** [vsg-dev/VulkanSceneGraph#838] [Q&A] [answered]
Hello,

I'm trying to build and run MyFirstVsgApplication on a Raspberry Pi 4. 

I've compiled and installed VSG apparently successfully. 

When I try to build MyFirstVsgApplication, or any of the vsgExampls or vsgXChange I get the same error during link:

/usr/bin/ld: /usr/local/lib/arm-linux-gnueabihf/libvsg.a(RecordTraversal.cpp.o): in function `vsg::RecordTraversal::apply(vsg::PagedLOD const&)':
RecordTraversal.cpp:(.text+0x1734): undefined reference to `__atomic_load_8'
/usr/bin/ld: RecordTraversal.cpp:(.text+0x1794): undefined reference to `__atomic_load_8'
/usr/bin/ld: RecordTraversal.cpp:(.text+0x17f4): undefined reference to `__atomic_exchange_8'
/usr/bin/ld: RecordTraversal.cpp:(.text+0x1844): undefined reference to `__atomic_load_8'
/usr/bin/ld: RecordTraversal.cpp:(.text+0x1880): undefined reference to `__atomic_compare_exchange_8'
/usr/bin/ld: /usr/local/lib/arm-linux-gnueabihf/libvsg.a(DatabasePager.cpp.o): in function `vsg::DatabaseQueue::take_when_available()':
DatabasePager.cpp:(.text+0x1708): undefined reference to `__atomic_load_8'
/usr/bin/ld: DatabasePager.cpp:(.text+0x1724): undefined reference to `__atomic_load_8'

(etc)

GCC/G++ Version: `gcc (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110`
cmake version 3.18.4

Any idea ? 

I've tried to add set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic") on CMakeLists.txt but it doesn't change anything.

Thanks !

**Comment by brunorzn** [comment]
using target_link_libraries worked, in CMakeLists:
```
-target_link_libraries(myfirstvsgapplication vsg::vsg)
+if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+    target_link_libraries(myfirstvsgapplication vsg::vsg -latomic)
+else()
+    target_link_libraries(myfirstvsgapplication vsg::vsg)
+endif()
```

**Comment by robertosfield** [comment]
Others have been using Pi4's and haven't reported this problem, I'm hoping we'll get feedback on what compiler/options being used to see if there is a pattern behind it.

I would rather merge a change to VSG rather than have 3rd party applications add platform specific tweaks to keep things compiling.  CMake config support should be able to help us here.

**Comment by reedev** [comment]
Hi,

Fwiw, on Raspbian 11 64 bit, I built everything from master (the VSG and projects) and it builds fine. As a first test I ran 'vsgdraw -d' and that''s running fine. Next, I checked out MyFirstVsgApplication, ran cmake, build and ran it. All fine working out of the box.

I would suggest to apt update, upgrade, reboot and build everything again.

If you are using a different setup, please let us know and we can try it out too.

**Reply by brunorzn** [reply]
Hi,

I'm on a 32 bits system, which is probably the thing

**Comment by vsg-dev** [comment]
@brunorzn and @reedev, I have been reviewing the VSG's CMake config set up and the following block of CMake script in [VulkanSceneGra/src/vsg/CMakeLists.txt](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/CMakeLists.txt#L242) looks relevant:

~~~ cmake
# Check for std::atomic
if(NOT MSVC AND NOT ANDROID AND NOT APPLE)
  check_cxx_source_compiles("
    #include <atomic>
    std::atomic<int> a;
    int main() { return a; }"
    HAVE_CXX_ATOMIC_WITHOUT_LIB
  )
  if(NOT HAVE_CXX_ATOMIC_WITHOUT_LIB)
    find_library(CXX_ATOMIC_LIBRARIES NAMES atomic atomic.so.1 libatomic.so.1 REQUIRED)
    list(APPEND LIBRARIES ${CXX_ATOMIC_LIBRARIES})
  endif()
endif()
~~~

My best guess is that this test is passing on 32bit build but the VSG is then using 64bit atomics which aren't supported.  If this is correct then just changing to test a wider range of atomics would help.  I will create a branch of the VSG and experiment.

**Comment by vsg-dev** [comment]
@brunorzn I have added a check for 64bit atomic to the check_cxx_source_compiles(..) test.

https://github.com/vsg-dev/VulkanSceneGraph/tree/atomic

The cmake script now looks like:
~~~ cmake 
# Check for std::atomic
if(NOT MSVC AND NOT ANDROID AND NOT APPLE)
  check_cxx_source_compiles("
    #include <atomic>
    std::atomic_uint32_t a32;
    std::atomic_uint64_t a64;
    int main() { return a32; }"
    HAVE_CXX_ATOMIC_WITHOUT_LIB
  )
  if(NOT HAVE_CXX_ATOMIC_WITHOUT_LIB)
    find_library(CXX_ATOMIC_LIBRARIES NAMES atomic atomic.so.1 libatomic.so.1 REQUIRED)
    list(APPEND LIBRARIES ${CXX_ATOMIC_LIBRARIES})
  endif()
endif()
~~~

@brunorzn  Could you test this out on your system, you'll need to revert your changes to MyFirstVsgApplication to make sure that things are now working out the box.  If this works I'll merged with master and it'll be part of the next VSG release.

**Comment by vsg-dev** [comment]
Hi @brunorzn could you test the atomic branch of the VSG, I've made a fix but can't merge with this with VSG master without confirmation that it works on Pi4 with 32bit.

**Comment by brunorzn** [comment]
Hi,

Sorry for my late reply, I was away from the PI for some time, back yesterday.

Unfortunately, I can't reproduce the problem anymore. I've cloned a fresh vsgExample and I can link it successfully on my PI. 

I'm looking at what may have changed - I've installed boost in the meantime, but I don't think that's it.

I'm investigating

**Comment by brunorzn** [comment]
I've tried with vsgXchange and was able to reproduce the error.

Adding your code didn't change anything, same error.

I had to add the following line also: +INCLUDE(CheckCXXSourceCompiles)

Adding the line target_link_libraries(vsgXchange INTERFACE -latomic) still fixes the error

Hope that helps

**Comment by robertosfield** [comment]
Where did you add the INCLUDE?  You provide a diff?

What error occurred if you didn't include the above change?

The adding of atmomic check in cmake is to see if -latomic is required then if it is then add it to link line.  It might be that the check is failing or the atomic lib isn't being added to the VSG's cmake config files that are installed in lib/cmake/vsg.

Unfortunately I can't test out and answer these questions as I don't have a Raspberry Pi4  so need the community that have them to step up and figure issues out.

**Comment by vsg-dev** [comment]
@brunorzn  I have just checked and the [VulkanSceneGraph/src/CMAkeLists.txt](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/CMakeLists.txt#L1) already has a INCLUDE(CheckCXXSourceCompiles) so I'm lost on where you may have needed to add this and why.

**Comment by brunorzn** [comment]
Apologies if I wasn't clear. 

VSG compiles without problem, it's when I compile vsgXChange that I got the problem.
I think VSGXChange's CMakeLists does not include CheckCXXSourceCompiles so I guess that why I've added this line.

I've tried on a fresh Raspbian 64 bits install, and I can confirm that everything works properly.

**Reply by vsg-dev** [reply]
What error do you see in vsgXchange when it doesn't have the CheckCXXSourceCompiles?

There isn't any check_cxx_source_compiles(..) usage in vsgXchange so I can't figure out why it would be needed in vsgXchange, one should only ever need to have a **include(CheckCXXSourceCompiles)** when the CMake script uses check_cxx_source_compiles.

---

FYI, when I force the VSG to include the atomic lib the install CMake config files vsgTargets.cmake include mention of the atomic lib:

```
ls ~/Install/lib/cmake/vsg$ grep atom *
vsgTargets.cmake:  INTERFACE_LINK_LIBRARIES "Vulkan::Vulkan;Threads::Threads;/usr/lib/x86_64-linux-gnu/libatomic.so.1;\$<LINK_ONLY:PkgConfig::xcb>"
```

When libs + applications then use the VSG's CMake config support to pull in the VSG these libs should be added automatically so there should be no need for these libs + applications to explicitly add them themselves.

**Reply by brunorzn** [reply]
I misunderstood that before. Let me check it out.

**Comment by vsg-dev** [comment]
I have decided to go ahead an merged my changes in the atomic branch as I think they generally make sense. #850.

**Comment by robertosfield** [comment]
On Tue, 13 Jun 2023 at 10:36, Bruno RZN ***@***.***> wrote:

> I misunderstood that before. Let me check it out.
>
> You should be able to use VSG master now.  I'll also be making a VSG-1.0.7
soon to include the change to atomic check.

**Reply by brunorzn** [reply]
I'll pull the change and test it out

**Reply by brunorzn** [reply]
It doesn't work. 

The code snippet seems to compile and thus HAVE_CXX_ATOMIC_WITHOUT_LIB is TRUE 

But still have those undefined references in vsgXchange:

__atomic_load_8
__atomic_exchange_8
__atomic_compare_exchange_8

**Reply by vsg-dev** [reply]
HAVE_CXX_ATOMIC_WITHOUT_LIB being TRUE will be the root of the problem.  At this point I don't know why the   check_cxx_source_compiles("
    #include <atomic>
    std::atomic_uint32_t a32;
    std::atomic_uint64_t a64;
    int main() { return a64; }"
    HAVE_CXX_ATOMIC_WITHOUT_LIB
  )

Should succeed by the later vsgXchange link fails.  Could it be the check_cxx_source_compiles(..) call only compiles, and doesn't link the compiled source?

**Reply by vsg-dev** [reply]
https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html

Suggests "Check if given C++ source compiles and links into an executable." so in theory it should be compiling and linking.

@brunorzn Could you just just forcing HAVE_CXX_ATOMIC_WITHOUT_LIB to FALSE before the line 

if(NOT HAVE_CXX_ATOMIC_WITHOUT_LIB)
    find_library(CXX_ATOMIC_LIBRARIES NAMES atomic atomic.so.1 libatomic.so.1 REQUIRED)
    list(APPEND LIBRARIES ${CXX_ATOMIC_LIBRARIES})
  endif()

Then rebuild and install the VSG then see if that can get vsgXchange to build cleanly.  This isn't a solution but might help provide insight into whether something in this direction might work.

**Reply by brunorzn** [reply]
forcing HAVE_CXX_ATOMIC_WITHOUT_LIB to FALSE works.

I tried to find a code that would compile on my laptop but not on my PI but didn't find it. 

My idea was to use something similar to what's written in vsg/threading/atomic.h but I wasn't able to have it compile anywhere with check_cxx_source_compiles.

**Reply by vsg-dev** [reply, marked as answer]
Putting more sophisticated code into check_cxx_source_compiles(..) might be able trigger an compile issue.  While we can't include vsg/threading/atomics.h we can recreate some of the key code that appears to be tripping up the normal compile.  I checked in to VSG master some addition atomic code inspired by include/vsg/threading/atomics.h: fce7e6063129218e94306baf1327a89bd0807da4

```
  check_cxx_source_compiles("
    #include <atomic>
    int main()
    {
        std::atomic_uint64_t a64(1);
        auto orig_value = a64.load();

        uint64_t new_value = 1;
        auto r = a64.compare_exchange_weak(orig_value, new_value);

        return 0;
    }"
    HAVE_CXX_ATOMIC_WITHOUT_LIB
  )
```

Could you try out the latest VSG master and see whether this fails the test?

**Reply by brunorzn** [reply]
That works ! Thank you :)

**Comment by brunorzn** [comment]
For the record, if anyone wants to have VSG running on a Raspberry Pi 4 running Raspbian 32 bits (as of june 2023), you just have to:

- burn the latest Raspbian 32 bits image, boot / configure the PI
- install the following packages: libxcb-xrm-dev cmake mesa-vulkan-drivers vulkan-tools libvulkan-dev

`sudo apt-get install libxcb-xrm-dev cmake mesa-vulkan-drivers vulkan-tools libvulkan-dev`

- follow VSG Readme, but don't download LunarG SDK

That's it ! You should be able to run the examples or MyFirstVSGApplication directly afterwards.

**Comment by vsg-dev** [comment]
Glad to hear it now works out of the box.  These fixes will be part of the VSG-1.0.7 release I'll tag later this week.
